extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -61,23 +70,65 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
+                @Override
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    // we could make the max number of entries configurable
+                    return size() > 30;
+                }
+            });
+
     public Uni<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
             return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
                     .map(new Function<SecurityIdentity, SecurityIdentity>() {
                         @Override
-                        public SecurityIdentity apply(SecurityIdentity securityIdentity) {
-                            return augmentIdentity(securityIdentity, tokens[1], tokens[2], context);
+                        public SecurityIdentity apply(SecurityIdentity identity) {
+                            if (isLogout(context, configContext)) {
+                                redirectToLogoutEndpoint(context, configContext, idToken);
+                            }
+
+                            return augmentIdentity(identity, accessToken, refreshToken, context);
+                        }
+                    }).on().failure().recoverWithItem(new Function<Throwable, SecurityIdentity>() {
+                        @Override
+                        public SecurityIdentity apply(Throwable throwable) {
+                            Throwable cause = throwable.getCause();
+
+                            // we should have proper exception hierarchy to represent token expiration errors
+                            if (cause != null && !cause.getMessage().equalsIgnoreCase("expired token")) {
+                                throw new AuthenticationFailedException(throwable);
+                            }
+
+                            SecurityIdentity identity = null;
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (isLogout(context, configContext) ||
+                                    (identity = trySilentRefresh(configContext, idToken, refreshToken, context,
+                                            identityProviderManager)) == null) {
+                                redirectToLogoutEndpoint(context, configContext, idToken);
```

Comment:
@pedroigor I think this looks nice but may be something still needs to be tweaked. So if we have a failure due to the expiry then 
1) if it is a logout request - we just redirect to the OP logout URI - this makes sense 
2) otherwise we just refresh, if identity is not null, all is good,  but if it is null, I think we should return AuthenticationFailedException so that a user's session is invalidated and we redirect the user to the login site, as opposed to the logout site, this is because the OP logout endpoint may not be even available, and even if it is available, I don't think the user, after his ID token has failed to refresh, should be redirected to the OP logout site - what if OP asks, would you like to logout ? while the users was doing some `GET /mybook`...

What do you think ?

Reply:
#1. We only redirect if the refresh fails. Otherwise, an identity is returned and the user is silently re-authenticated.

#2. You are right, this is wrong. We should restart the authentication.